# Aktivitas partisipatif #3 (RandomForestClassifier)

Nama: Muhammad Karov Ardava Barus

NIM: 103052300001

Kelas: DS-47-01

## Import Libraries

Di tahap ini, kita akan memanggil libraries yang akan digunakan untuk membersihkan dataset yang kita dapatin

In [15]:
import pandas as pd # manipulasi data tabular
from sklearn.preprocessing import MinMaxScaler, LabelEncoder # preprocessing data
from sklearn.model_selection import train_test_split # pemisahan data
from sklearn.naive_bayes import GaussianNB # model Naive Bayes
from sklearn.ensemble import RandomForestClassifier # model Random Forest
from sklearn.metrics import accuracy_score, confusion_matrix, f1_score # evaluasi model

**a. Memanggil Data**

In [16]:
df = pd.read_excel('CAD alizadeh.xls') # memanggil data dengan format excel (.xls, .xlsx)

**b. Menampilkan data**

In [17]:
df.head(10) # menampilkan 10 data teratas

No  Age  Weight  Length    Sex        BMI  DM  HTN  Current Smoker  \
0   1   53      90     175   Male  29.387755   0    1               1   
1   2   67      70     157  Fmale  28.398718   0    1               0   
2   3   54      54     164   Male  20.077335   0    0               1   
3   4   66      67     158  Fmale  26.838648   0    1               0   
4   5   50      87     153  Fmale  37.165193   0    1               0   
5   6   50      75     175   Male  24.489796   0    0               1   
6   7   55      80     165   Male  29.384757   0    0               0   
7   8   72      80     175   Male  26.122449   1    0               1   
8   9   58      84     163  Fmale  31.615793   0    0               0   
9  10   60      71     170   Male  24.567474   1    0               0   

   EX-Smoker  ...    K   Na    WBC Lymph Neut  PLT EF-TTE Region RWMA     VHD  \
0          0  ...  4.7  141   5700    39   52  261     50           0       N   
1          0  ...  4.7  156   7700    38   55  165     40           4       N   
2          0  ...  4.7  139   7400    38   60  230     40           2    mild   
3          0  ...  4.4  142  13000    18   72  742     55           0  Severe   
4          0  ...  4.0  140   9200    55   39  274     50           0  Severe   
5          0  ...  4.2  141   7300    26   66  194     50           0       N   
6          1  ...  4.8  139   9400    58   33  292     40           4    mild   
7          0  ...  4.3  142  12200    25   74  410     45           4    mild   
8          0  ...  3.4  139   5100    49   50  370     50           0       N   
9          0  ...  4.6  140   4900    55   42  380     40           2       N   

     Cath  
0     Cad  
1     Cad  
2     Cad  
3  Normal  
4  Normal  
5     Cad  
6     Cad  
7     Cad  
8  Normal  
9     Cad  

[10 rows x 56 columns]

## Data Preprocessing

**c. Memeriksa apakah ada yang kosong pada data**

In [18]:
missing_values = df.isna().sum().sort_values(ascending=False) # mengecek missing value 

print('Total missing values tiap kolom:', '\n', missing_values) # menampilkan total missing value

Total missing values tiap kolom: 
 No                       0
Age                      0
Exertional CP            0
LowTH Ang                0
Q Wave                   0
St Elevation             0
St Depression            0
Tinversion               0
LVH                      0
Poor R Progression       0
FBS                      0
CR                       0
TG                       0
LDL                      0
HDL                      0
BUN                      0
ESR                      0
HB                       0
K                        0
Na                       0
WBC                      0
Lymph                    0
Neut                     0
PLT                      0
EF-TTE                   0
Region RWMA              0
VHD                      0
Nonanginal               0
Atypical                 0
Function Class           0
CVA                      0
Weight                   0
Length                   0
Sex                      0
BMI                      0
DM                  

In [19]:
missing_values = df.isna().sum().sum() # mengecek total missing value 

print(f'Total missing values: {missing_values}') # menampilkan total missing value

Total missing values: 0


In [20]:
print(f'Total data duplikat: {df.duplicated().sum()}') # mengecek total data duplikat

Total data duplikat: 0


**d. Mengubah semua kolom menjadi bernilai numerik**

Untuk mengubah data kategorik menjadi data numerik, kita akan menggunakan metode `Label Encoding`

In [21]:
cat_cols = df.select_dtypes(include='object').columns # mengambil kolom kategorikal
num_cols = df.select_dtypes(include='number').columns # mengambil kolom numerik

In [22]:
le = LabelEncoder() # inisialisasi label encoder

for col in cat_cols: # loop untuk setiap kolom kategorikal
    df[col] = le.fit_transform(df[col]) # transformasi data kategorikal menjadi numerik

df.head(10) # menampilkan 10 data teratas

No  Age  Weight  Length  Sex        BMI  DM  HTN  Current Smoker  \
0   1   53      90     175    1  29.387755   0    1               1   
1   2   67      70     157    0  28.398718   0    1               0   
2   3   54      54     164    1  20.077335   0    0               1   
3   4   66      67     158    0  26.838648   0    1               0   
4   5   50      87     153    0  37.165193   0    1               0   
5   6   50      75     175    1  24.489796   0    0               1   
6   7   55      80     165    1  29.384757   0    0               0   
7   8   72      80     175    1  26.122449   1    0               1   
8   9   58      84     163    0  31.615793   0    0               0   
9  10   60      71     170    1  24.567474   1    0               0   

   EX-Smoker  ...    K   Na    WBC  Lymph  Neut  PLT  EF-TTE  Region RWMA  \
0          0  ...  4.7  141   5700     39    52  261      50            0   
1          0  ...  4.7  156   7700     38    55  165      40            4   
2          0  ...  4.7  139   7400     38    60  230      40            2   
3          0  ...  4.4  142  13000     18    72  742      55            0   
4          0  ...  4.0  140   9200     55    39  274      50            0   
5          0  ...  4.2  141   7300     26    66  194      50            0   
6          1  ...  4.8  139   9400     58    33  292      40            4   
7          0  ...  4.3  142  12200     25    74  410      45            4   
8          0  ...  3.4  139   5100     49    50  370      50            0   
9          0  ...  4.6  140   4900     55    42  380      40            2   

   VHD  Cath  
0    1     0  
1    1     0  
2    3     0  
3    2     1  
4    2     1  
5    1     0  
6    3     0  
7    3     0  
8    1     1  
9    1     0  

[10 rows x 56 columns]

**e. Melakukan proses normalisasi pada masing-masing atribut**

Normalisasi biasanya diterapkan pada data numerik untuk memastikan bahwa semua fitur berada dalam skala yang sama, tetapi **tidak perlu** dinormalisasiin pada data kategorikal. Metode yang kita gunakan adalah `Min-Max Scaling`

In [23]:
scaler = MinMaxScaler() # inisialisasi minmax scaler

df[num_cols] = scaler.fit_transform(df[num_cols]) # transformasi data numerik menggunakan minmax scaler

df.head(10) # menampilkan 10 data teratas

No       Age    Weight    Length  Sex       BMI   DM  HTN  \
0  0.000000  0.410714  0.583333  0.729167    1  0.494721  0.0  1.0   
1  0.003311  0.660714  0.305556  0.354167    0  0.451314  0.0  1.0   
2  0.006623  0.428571  0.083333  0.500000    1  0.086105  0.0  0.0   
3  0.009934  0.642857  0.263889  0.375000    0  0.382846  0.0  1.0   
4  0.013245  0.357143  0.541667  0.270833    0  0.836058  0.0  1.0   
5  0.016556  0.357143  0.375000  0.729167    1  0.279759  0.0  0.0   
6  0.019868  0.446429  0.444444  0.520833    1  0.494590  0.0  0.0   
7  0.023179  0.750000  0.444444  0.729167    1  0.351413  1.0  0.0   
8  0.026490  0.500000  0.500000  0.479167    0  0.592505  0.0  0.0   
9  0.029801  0.535714  0.319444  0.625000    1  0.283168  1.0  0.0   

   Current Smoker  EX-Smoker  ...         K        Na       WBC     Lymph  \
0             1.0        0.0  ...  0.472222  0.464286  0.139860  0.603774   
1             0.0        0.0  ...  0.472222  1.000000  0.279720  0.584906   
2             1.0        0.0  ...  0.472222  0.392857  0.258741  0.584906   
3             0.0        0.0  ...  0.388889  0.500000  0.650350  0.207547   
4             0.0        0.0  ...  0.277778  0.428571  0.384615  0.905660   
5             1.0        0.0  ...  0.333333  0.464286  0.251748  0.358491   
6             0.0        1.0  ...  0.500000  0.392857  0.398601  0.962264   
7             1.0        0.0  ...  0.361111  0.500000  0.594406  0.339623   
8             0.0        0.0  ...  0.111111  0.392857  0.097902  0.792453   
9             0.0        0.0  ...  0.444444  0.428571  0.083916  0.905660   

       Neut       PLT    EF-TTE  Region RWMA  VHD  Cath  
0  0.350877  0.329149  0.777778          0.0    1     0  
1  0.403509  0.195258  0.555556          1.0    1     0  
2  0.491228  0.285914  0.555556          0.5    3     0  
3  0.701754  1.000000  0.888889          0.0    2     1  
4  0.122807  0.347280  0.777778          0.0    2     1  
5  0.596491  0.235704  0.777778          0.0    1     0  
6  0.017544  0.372385  0.555556          1.0    3     0  
7  0.736842  0.536960  0.666667          1.0    3     0  
8  0.315789  0.481172  0.777778          0.0    1     1  
9  0.175439  0.495119  0.555556          0.5    1     0  

[10 rows x 56 columns]

**⚠️ Tambahan: Di feature `No` sebenarnya tidak bermakna, karena hanya sebagai index, contoh: (1, 2, 3, ..., n), sehingga disarankan untuk dihapus saja.**

In [24]:
df.drop(columns=['No'], inplace=True) # menghapus kolom No

df.head(10) # menampilkan 10 data teratas

Age    Weight    Length  Sex       BMI   DM  HTN  Current Smoker  \
0  0.410714  0.583333  0.729167    1  0.494721  0.0  1.0             1.0   
1  0.660714  0.305556  0.354167    0  0.451314  0.0  1.0             0.0   
2  0.428571  0.083333  0.500000    1  0.086105  0.0  0.0             1.0   
3  0.642857  0.263889  0.375000    0  0.382846  0.0  1.0             0.0   
4  0.357143  0.541667  0.270833    0  0.836058  0.0  1.0             0.0   
5  0.357143  0.375000  0.729167    1  0.279759  0.0  0.0             1.0   
6  0.446429  0.444444  0.520833    1  0.494590  0.0  0.0             0.0   
7  0.750000  0.444444  0.729167    1  0.351413  1.0  0.0             1.0   
8  0.500000  0.500000  0.479167    0  0.592505  0.0  0.0             0.0   
9  0.535714  0.319444  0.625000    1  0.283168  1.0  0.0             0.0   

   EX-Smoker   FH  ...         K        Na       WBC     Lymph      Neut  \
0        0.0  0.0  ...  0.472222  0.464286  0.139860  0.603774  0.350877   
1        0.0  0.0  ...  0.472222  1.000000  0.279720  0.584906  0.403509   
2        0.0  0.0  ...  0.472222  0.392857  0.258741  0.584906  0.491228   
3        0.0  0.0  ...  0.388889  0.500000  0.650350  0.207547  0.701754   
4        0.0  0.0  ...  0.277778  0.428571  0.384615  0.905660  0.122807   
5        0.0  0.0  ...  0.333333  0.464286  0.251748  0.358491  0.596491   
6        1.0  0.0  ...  0.500000  0.392857  0.398601  0.962264  0.017544   
7        0.0  0.0  ...  0.361111  0.500000  0.594406  0.339623  0.736842   
8        0.0  0.0  ...  0.111111  0.392857  0.097902  0.792453  0.315789   
9        0.0  0.0  ...  0.444444  0.428571  0.083916  0.905660  0.175439   

        PLT    EF-TTE  Region RWMA  VHD  Cath  
0  0.329149  0.777778          0.0    1     0  
1  0.195258  0.555556          1.0    1     0  
2  0.285914  0.555556          0.5    3     0  
3  1.000000  0.888889          0.0    2     1  
4  0.347280  0.777778          0.0    2     1  
5  0.235704  0.777778          0.0    1     0  
6  0.372385  0.555556          1.0    3     0  
7  0.536960  0.666667          1.0    3     0  
8  0.481172  0.777778          0.0    1     1  
9  0.495119  0.555556          0.5    1     0  

[10 rows x 55 columns]

Data kita sudah bersih!

## Modelling
**Split data menjadi training dan test set**


Kita akan split data train dan test menjadi 80:20

In [25]:
# membagi data menjadi data train dan data test
X_train, X_test, y_train, y_test = train_test_split(
    df.drop(columns=['Cath']), # features
    df['Cath'], # target (yang ingin diprediksi)
    test_size=0.3, # ukuran data test 20% dari total data
    random_state=42 # seed agar hasil random tetap konsisten
)

**Pemodelan dengan Naive Bayes**

In [ ]:
model_nb = GaussianNB() # inisialisasi model Naive Bayes
model_nb.fit(X_train, y_train) # training model

y_pred_nb = model_nb.predict(X_test) # prediksi data test

print(f'Akurasi model: {accuracy_score(y_test, y_pred_nb)}') # evaluasi model menggunakan accuracy score
print(f'F1 score model: {f1_score(y_test, y_pred_nb)}') # evaluasi model menggunakan f1 score
print(f'Confusion matrix model: \n{confusion_matrix(y_test, y_pred_nb)}') # evaluasi model menggunakan confusion matrix

Akurasi model: 0.4945054945054945
F1 score model: 0.5208333333333334
Confusion matrix model: 
[[20 46]
 [ 0 25]]


**Pemodelan dengan Random Forest**

In [27]:
model_rf = RandomForestClassifier(n_estimators=100, random_state=42) # inisialisasi model Random Forest
model_rf.fit(X_train, y_train) # training model

y_pred_rf = model_rf.predict(X_test) # prediksi data test

print(f'Akurasi model: {accuracy_score(y_test, y_pred_rf)}') # evaluasi model menggunakan accuracy score
print(f'F1 score model: {f1_score(y_test, y_pred_rf)}') # evaluasi model menggunakan f1 score
print(f'Confusion matrix model: \n{confusion_matrix(y_test, y_pred_rf)}') # evaluasi model menggunakan confusion matrix

Akurasi model: 0.8571428571428571
F1 score model: 0.6976744186046512
Confusion matrix model: 
[[63  3]
 [10 15]]


**Hasil analisis**

Dari hasil yang ditampilkan, terlihat jelas bahwa model Random Forest memberikan performa yang jauh lebih baik dibandingkan Naive Bayes. Akurasi Random Forest mencapai sekitar 85.7%, sedangkan Naive Bayes hanya sekitar 49.4%, yang bahkan mendekati nilai tebakan acak jika kelasnya seimbang. Selain itu, F1 score dari Random Forest juga jauh lebih tinggi (sekitar 0.69 dibandingkan 0.52), yang menunjukkan bahwa Random Forest mampu menangkap keseimbangan antara precision dan recall dengan lebih baik.

Kalau kita lihat dari confusion matrix-nya, performa Naive Bayes sangat ga balanced, model ini cenderung memprediksi kelas 1 secara berlebihan, sampai-sampai tidak ada satu pun kasus dari kelas 1 yang diprediksi sebagai kelas 0 (false negative-nya nol), tapi justru menghasilkan sangat banyak false positive (46 kasus). Ini menunjukkan bahwa asumsi independensi antar fitur pada Naive Bayes tampaknya tidak terpenuhi dalam dataset ini, sehingga membuat prediksi menjadi kurang akurat.

Di sisi lain, Random Forest tampil lebih seimbang. False positive dan false negative-nya relatif rendah, dan model ini berhasil mengklasifikasikan sebagian besar data dengan benar, baik dari kelas 0 maupun kelas 1. Ini tidak mengherankan, karena Random Forest adalah model yang jauh lebih kompleks dan fleksibel dalam menangani data yang memiliki relasi non-linear antar fitur.

Secara keseluruhan, dari analisis ini bisa disimpulkan bahwa Random Forest merupakan pilihan yang jauh lebih tepat untuk dataset ini dibandingkan Naive Bayes, terutama jika tujuannya adalah untuk mendapatkan model yang andal dan stabil dalam melakukan klasifikasi.